pandas/tests/groupby/methods/test_value_counts.py

```diff
@@ -1177,3 +1177,32 @@ def test_value_counts_integer_columns():
         {1: ["a", "a", "a"], 2: ["a", "a", "d"], 3: ["a", "b", "c"], "count": 1}
     )
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("vc_sort", [True, False])
+@pytest.mark.parametrize("normalize", [True, False])
+def test_value_counts_sort(sort, vc_sort, normalize):
+    # GH#55951
+    df = DataFrame({"a": [2, 1, 1, 1], 0: [3, 4, 3, 3]})
```

Comment:
Could you also tests with `Categorical`?

Reply:
Thanks - good call. I added a test and found a bug calling `.sortlevel()` in the categorical section. But the categorical behavior is still buggy here; namely 

https://github.com/pandas-dev/pandas/blob/ef2c61ae69add5c6fb960cd235b10b0e741b9aa4/pandas/core/groupby/groupby.py#L2824-L2826

is constructing the index by taking the cartesian product across all groupings; it should only be doing this for the categorical variables. That's not an easy fix, but will be handled in #55738 where we can just remove that whole block of code.